streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -272,15 +280,22 @@ public synchronized void onChange(final StreamThread thread,
             }
             if (newState == StreamThread.State.PARTITIONS_REVOKED ||
                 newState == StreamThread.State.ASSIGNING_PARTITIONS) {
-                setState(State.REBALANCING);
+                setStateWhenNotShuttingDownOrNotRunning(State.REBALANCING);
```

Comment:
Ditto as below, let's think through how we want to set the state of the instance if the following can ever happen: some of the threads are shutting down or is already shut down, while others are continue running. If yes how we want to set the state.